### PR TITLE
Propagate errors when testing

### PR DESCRIPTION
**Pull request by smurfix**
Testing with Mocha requires errors raised within flows to propagate to the test runner, not getting swallowed by Node-Red.

This commit tests whether Mocha's global functions exist and, if they do, propagates exceptions that are raised there.

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Proposed changes

When running under Mocha, Node-Red must propagate exceptions to the test runner so that suitable error messages can be emitted instead of a timeout.

This closes issue #4956.

Suggestions how to test for this feature, should such a test be deemed necessary, are welcome. I have to admit that I am not well-versed enough in Javascript, Mocha and/or Node-Red lore to write a test for this specific issue that's not (a) visual, or (b) would be a copy of the test that the "detect-mocha" module already does.

## Checklist

- [X] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [ ] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [X] I have run `npm run test` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality

### CODE REVIEW

**Comment by dceejay** [comment]
Do we really need to include a 5 line npm dependency to do this ? (Which itself depends on a 1 line package...)

**Comment by smurfix** [comment]
I can easily vendorize the package if that's more palatable. No problem.

**Comment by knolleary** [comment]
I'm not a fan of changing behaviour of the core depending on how its being run. There may be existing tests that rely on this behaviour and actively check the error log output to detect errors - this would break those tests.

It may be cleaner for Test Helper to validate the log has no errors in it automatically and to report up any issues.

**Comment by dceejay** [comment]
Quite a few nodes already have try/catch around individual tests which fixes this in those specific cases so what is returned may not always be the timeout and thus (as pointed out) returns better error messages. If there is another way of fixing the testing framework such that it does this without having to rewrite most of the node tests individually that would I think be a benefit.

**Comment by smurfix** [comment]
> There may be existing tests that rely on this behaviour

I respectfully submit that that's not valid argument for declaring this behavior a feature.

Discovering how to fix this is far from trivial. Even finding the root cause required me to spend multiple hours of learning how to debug a framework which I'd very much like to not step on my toes in the first place.

IMHO adding such a workaround to a testcase requires a far deeper understanding of Javascript, Node, and Mocha: how to add tests is documented fairly well in multiple places — but how to add a test that doesn't time out when it fails … not so much.
Also, you can't do that retroactively, thus when a test starts failing in CI (e.g. due to `dependabot` updating something) you're *still* SOL and, more likely than not, need multiple iterations to even discover *where* to add the workaround.

In other words, applying this patch (or something like it) and then removing those workarounds from the packages that currently require them seems to be a rather more effective use of all our development time.

NB: At first glance, an alternate solution would be to change the core so that only Mocha's exceptions are propagated. I rejected this approach because it doesn't help at all when the test raises an unrelated exception, such as something trying to call (or access an attribute of) `undefined`, which is far more likely than we'd all want it to be.

**Comment by dceejay** [comment]
I tend to agree - we can't be too far away from v5 (as node20 is now into Maint mode) - and as we do release betas that should give people time to test for any breakage in tests. But I still think it must be possible to detect test mode without using those libs (EG duplicating those 6 line :-) - or maybe using an ENV variable instead ?

**Comment by smurfix** [comment]
Using an envvar or similar makes sense _if_ we can convince the test runner to set it by default.

Otherwise testing for Mocha's globals seems like an acceptable substitute, they're not going to change them after all; vendoring these two admittedly-micro-libs is, umm, not exactly rocket science. ;-)

**Comment by knolleary** [comment]
> IMHO adding such a workaround to a testcase...

I was possibly not clear in my reply. I'm not saying individual test cases need to workaround this. I am suggesting looking at how `node-red-node-test-helper` could help here as a universal solution. This is the standard test framework all node tests use. It could monitor the error log and trigger a failure if an error is logged by the runtime (although off the top of my head, not 100% sure how that would work...)

My point was I'd like to evaluate these options before going down the path of adding conditional behaviour in the core runtime.

> Using an envvar or similar makes sense if we can convince the test runner to set it by default.

If we go down the conditional behaviour route, given all tests are run via `node-red-node-test-helper`, then it could be updated to set whatever env var we want.
